### DISCUSSION TITLE
What is the max sampling rate?

### DISCUSSION THREAD

**Discussion by mattster98** [emporia-vue-local/esphome#294] [Q&A] [answered]
I don't yet have an emporia yet but am looking for an additional monitoring for my grid input to the house because my existing monitoring is now behind my solar/inverter/battery system so I don't have the direct grid visibility I used to.  The inverters give 10 second interval data but I had an event this morning that I can't quite figure out so I'm looking to add dedicated, higher resolution monitoring of the grid inputs.  I'm not looking for Fluke-level waveforms, but sub-second resolution at least on the voltage/frequency would be helpful.

**Comment by gekkehenkie11** [comment, marked as answer]
Currently the Atmel chip that does the sampling, gives 2 times per second an update to the ESP chip. However I've made the Atmel firmware open source: https://github.com/gekkehenkie11/emporia-SAMD09 So if you can program, you could update the code to increase that update frequency.

**Reply by mattster98** [reply]
Excellent!  I guess the theoretical limit is some combination of the samples/sec of the SAMD09 and the bandwidth of the connection between it and the ESP.  Curious which is more likely to be the bottleneck.

**Reply by gekkehenkie11** [reply]
yeah exactly. I suspect that the I2C connection between the chips is the bottle neck. But if you want to have a packet to the ESP 200 times per second for example, I don't think that should be a problem. But if you want to send each new sample to the ESP once it's ready, I doubt it will be able to keep up. Currently about 25.000 samples/sec if I remember correctly.

**Reply by mattster98** [reply]
Yeah, maybe a "power quality" mode might be in order to sample somehow and detect phase anomalies or sudden frequency shifts then send a "bulk download" of the event data that is high resolution vs always sending that resolution.  I'll be ordering one soon to start tinkering. :D

Is the SAMD09 firmware update built into the main emporia-vue-local or is it a separate step?  Feel free to point me to docs if this is a FAQ.. haven't run across those details.

**Reply by gekkehenkie11** [reply]
It's a separate step, see my readme: https://github.com/gekkehenkie11/Emporia-VUE-fix/blob/main/README.md
It's also possible to flash it via the ESP, @odewdney wrote something for it: https://github.com/odewdney/emporiavue but he never explained it how it works and I have never really looked into that route myself. But if someone figures that out, you wouldn't need to use a JTAG adaptor anymore to flash the Atmel chip.

**Comment by rct** [comment]
I would also be very interested in any sort of features, a mode, or separate firmware that could be used to detect anomalies/power quality issues.  I'd like to be able to check:

* solar inverters are working correctly and generating "clean" power
* backup generator is working correctly and generating "clean" power
* There aren't transient problems with the utility power coming from the pole.   (In the area I'm in, the couplers for the riser cables and and some closer to the transformer corrode and need to be replaced every few years when the lights start flickering.)

**Reply by mattster98** [reply]
I hadn't even considered the generator power quality. That's a good one.  This whole question started from a very brief utility power event yesterday that took my inverters offline for a couple minutes.  My monitoring sample rate was insufficient to see anything more than a couple blips on the AC frequency but that shouldn't have triggered what happened.
